lib/kino/js/live.ex

```diff
@@ -0,0 +1,369 @@
+      ctx.handleEvent("update", (payload) => {
+        // ...
+      });
+    });
+  }
+  ```
+
+  Or alternatively:
+
+  ```js
+  export async function init(ctx, data) {
+    const response = await fetch(data.someUrl);
+
+    ctx.handleEvent("update", (payload) => {
+      // ...
+    });
+  }
+  ```
+
+  In such case all incoming events are buffered and dispatched once
+  the handler is registered.
+  '''
+
+  defstruct [:module, :pid]
+
+  @doc false
+  use GenServer
+
+  require Logger
+
+  alias Kino.JS.Live.Context
+
+  @type t :: %__MODULE__{module: module(), pid: pid()}
+
+  @doc """
+  Invoked when the widget server started.
+
+  See `c:GenServer.init/1` for more details.
+  """
+  @callback init(arg :: term(), ctx :: Context.t()) :: {:ok, ctx :: Context.t()}
+
+  @doc """
+  Invoked whenever a new client connects to the server.
+
+  The returned data is passed to the JavaScript `init` function
+  of the connecting client.
+  """
+  @callback handle_connect(ctx :: Context.t()) :: {:ok, data :: term(), ctx :: Context.t()}
+
+  @doc """
+  Invoked to handle client events.
+  """
+  @callback handle_event(event :: String.t(), payload :: term(), ctx :: Context.t()) ::
+              {:noreply, ctx :: Context.t()}
+
+  @doc """
+  Invoked to handle asynchronous `cast/2` messages.
+
+  See `c:GenServer.handle_cast/2` for more details.
+  """
+  @callback handle_cast(msg :: term(), ctx :: Context.t()) :: {:noreply, ctx :: Context.t()}
+
+  @doc """
+  Invoked to handle synchronous `call/3` messages.
+
+  See `c:GenServer.handle_call/3` for more details.
+  """
+  @callback handle_call(msg :: term(), {pid(), reference()}, ctx :: Context.t()) ::
+              {:noreply, ctx :: Context.t()} | {:reply, term(), ctx :: Context.t()}
+
+  @doc """
+  Invoked to handle all other messages.
+
+  See `c:GenServer.handle_info/2` for more details.
+  """
+  @callback handle_info(msg :: term(), ctx :: Context.t()) :: {:noreply, ctx :: Context.t()}
+
+  @doc """
+  Invoked when the server is about to exit.
+
+  See `c:GenServer.terminate/2` for more details.
+  """
+  @callback terminate(reason, ctx :: Context.t()) :: term()
+            when reason: :normal | :shutdown | {:shutdown, term()}
+
+  @optional_callbacks init: 2,
+                      handle_event: 3,
+                      handle_call: 3,
+                      handle_cast: 2,
+                      handle_info: 2,
+                      terminate: 2
+
+  defmacro __using__(opts) do
+    quote location: :keep, bind_quoted: [opts: opts] do
+      @behaviour Kino.JS.Live
+
+      import Kino.JS.Live.Context, only: [assign: 2, update: 3, broadcast_event: 3]
+
+      @before_compile Kino.JS.Live
+    end
+  end
+
+  def __before_compile__(env) do
+    unless Module.defines?(env.module, {:__js_info__, 0}) do
+      message = """
+      make sure to include Kino.JS in #{inspect(env.module)} and define the necessary assets.
+
+          use Kino.JS
+          use Kino.JS.Live
```

Comment:
Is there any reason to tell them to `use Kino.JS.Live`? Might as well tell them to `use Kino.JS` but then start it with `Kino.JS.Live`. It is up to them to import the assign functions and friends.

Reply:
Fair point, I was trying to stress out that `Kino.JS.Live` doesn't automatically inject `Kino.JS`, but juse `use Kino.JS` sounds good to me!